### ISSUE TITLE
nodejs error: spawn ENOENT

### ISSUE THREAD

**Issue by miguelmota** [RocketChat/Rocket.Chat#10187]
### Description:

Get error `nodejs error: spawn ENOENT` when install modules

### Server Setup Information:

* Version of Rocket.Chat Server: 0.63.0-develop
* Operating System: Ubuntu 17.10
* Deployment Method(snap/docker/tar/etc): docker
* Number of Running Instances: 1
* DB Replicaset Oplog: 0
* Node Version: 9.5.0
* mongoDB Version: v3.4.10

### Steps to Reproduce:

1. cd /app/bundle/programs/server
2. npm install

### Expected behavior:

successful install

### Actual behavior:

```
> meteor-dev-bundle@ install /app/bundle/programs/server
> node npm-rebuild.js

events.js:165
      throw er; // Unhandled 'error' event
      ^

Error: spawn npm ENOENT
    at Process.ChildProcess._handle.onexit (internal/child_process.js:201:19)
    at onErrorNT (internal/child_process.js:379:16)
    at process._tickCallback (internal/process/next_tick.js:114:19)
    at Function.Module.runMain (module.js:692:11)
    at startup (bootstrap_node.js:194:16)
    at bootstrap_node.js:666:3
Emitted 'error' event at:
    at Process.ChildProcess._handle.onexit (internal/child_process.js:207:12)
    at onErrorNT (internal/child_process.js:379:16)
    [... lines matching original stack trace ...]
    at bootstrap_node.js:666:3
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! meteor-dev-bundle@ install: `node npm-rebuild.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the meteor-dev-bundle@ install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/.npm/_logs/2018-03-21T06_54_07_257Z-debug.log
The command '/bin/sh -c npm install' returned a non-zero code: 1
```

### Relevant logs:

```
Step 25/30 : RUN npm install
 ---> Running in 0c1e0e86515c

> fibers@2.0.0 install /app/bundle/programs/server/node_modules/fibers
> node build.js || nodejs build.js

make: Entering directory '/app/bundle/programs/server/node_modules/fibers/build'
  CXX(target) Release/obj.target/fibers/src/fibers.o
../src/fibers.cc: In function 'void uni::SetAccessor(v8::Isolate*, v8::Local<v8::Object>, v8::Local<v8::String>, uni::FunctionType (*)(v8::Local<v8::String>, const GetterCallbackInfo&), void (*)(v8::Local<v8::String>, v8::Local<v8::Value>, const SetterCallbackInfo&))':
../src/fibers.cc:318:129: warning: ignoring return value of 'v8::Maybe<bool> v8::Object::SetAccessor(v8::Local<v8::Context>, v8::Local<v8::Name>, v8::AccessorNameGetterCallback, v8::AccessorNameSetterCallback, v8::MaybeLocal<v8::Value>, v8::AccessControl, v8::PropertyAttribute)', declared with attribute warn_unused_result [-Wunused-result]
   object->SetAccessor(isolate->GetCurrentContext(), name, (AccessorNameGetterCallback)getter, (AccessorNameSetterCallback)setter);
                                                                                                                                 ^
In file included from /home/.node-gyp/9.8.0/include/node/node.h:63:0,
                 from ../src/coroutine.h:1,
                 from ../src/fibers.cc:1:
/home/.node-gyp/9.8.0/include/node/v8.h:3240:37: note: declared here
   V8_WARN_UNUSED_RESULT Maybe<bool> SetAccessor(Local<Context> context,
                                     ^~~~~~~~~~~
  CXX(target) Release/obj.target/fibers/src/coroutine.o
  CC(target) Release/obj.target/fibers/src/libcoro/coro.o
  SOLINK_MODULE(target) Release/obj.target/fibers.node
  COPY Release/fibers.node
make: Leaving directory '/app/bundle/programs/server/node_modules/fibers/build'
Installed in `/app/bundle/programs/server/node_modules/fibers/bin/linux-x64-59/fibers.node`

> meteor-dev-bundle@ install /app/bundle/programs/server
> node npm-rebuild.js

events.js:165
      throw er; // Unhandled 'error' event
      ^

Error: spawn npm ENOENT
    at Process.ChildProcess._handle.onexit (internal/child_process.js:201:19)
    at onErrorNT (internal/child_process.js:379:16)
    at process._tickCallback (internal/process/next_tick.js:114:19)
    at Function.Module.runMain (module.js:692:11)
    at startup (bootstrap_node.js:194:16)
    at bootstrap_node.js:666:3
Emitted 'error' event at:
    at Process.ChildProcess._handle.onexit (internal/child_process.js:207:12)
    at onErrorNT (internal/child_process.js:379:16)
    [... lines matching original stack trace ...]
    at bootstrap_node.js:666:3
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! meteor-dev-bundle@ install: `node npm-rebuild.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the meteor-dev-bundle@ install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/.npm/_logs/2018-03-21T06_54_07_257Z-debug.log
The command '/bin/sh -c npm install' returned a non-zero code: 1
```

**Comment by MartinSchoeler** [comment]
We aim to keep our Github issues for bugs and feature requests.  This does not fit that criteria.
Please see our [guidelines for support](https://rocket.chat/docs/getting-support).  If you feel like we are closing in error, please re-open and add additional details about the bug or feature request.

**Comment by KrinalPatel11** [comment]
NOTE: This error is almost always caused because the command does not exist, because the working directory does not exist, or from a windows-only bug. I found a particular easy way to get the idea of the root cause of:

Error: spawn ENOENT
The problem of this error is, there is really little information in the error message to tell you where the call site is, i.e. which executable/command is not found, especially when you have a large code base where there are a lot of spawn calls. I found a very easy way to spot which command cause the problem rather than adding event listeners everywhere in your code as suggested in @laconbass’ answer. The key idea is to wrap the original spawn call with a wrapper which prints the arguments send to the spawn call. Here is the wrapper function, put it at the top of the index.js or whatever your server’s starting script.

(function() {
var childProcess = require("child_process");
var oldSpawn = childProcess.spawn;
function mySpawn() {
console.log('spawn called');
console.log(arguments);
var result = oldSpawn.apply(this, arguments);
return result;
}
childProcess.spawn = mySpawn;
})();

Full Answer With Ex. [Here](https://icetutor.com/question/how-do-i-debug-error-spawn-enoent-on-node-js/)
